### ISSUE TITLE
[JIT] JIT erroneously sets requires grad

### ISSUE THREAD

**Issue by t-vi** [pytorch/pytorch#18270]
## 🐛 Bug

Erroneous `requires_grad` set.

## To Reproduce

I don't know if this is exactly the same bug as #17532 , it might be closely related
```
@torch.jit.script
def fn(x, y: Optional[torch.Tensor]=None, c:float=0.1):
  if y is not None:
      y.copy_(y * c + (1-c) * x.detach())
  return x

a = torch.randn(5,5, device='cuda', requires_grad=True)
b = torch.randn_like(a, requires_grad=False)
fn(a, b)
print (b.requires_grad)
```

## Expected behavior

`b` should not require grad.

## Environment

Master PyTorch

## Additional context

So this is similar to @ngimel's report in #17532 in the way that she has lists while I have Optional to unbox. Also for some reason I need the calculation bit, a mery `copy_(x)` would not show the bug.

**Comment by apaszke** [comment]
Why is it not supposed to require grad? You're copying a value that requires grad into its data, so it seems correct to me.

**Comment by t-vi** [comment]
No, I'm copying an expression of `x.detach()` and `y` (which does not require grad).
So the easy part is to fix detach to actually set requires grad to false for dimensioned tensors (sending a PR right now), but the thing is that Optional, like list, will only record "Tensor" as type, not DimensionedTensor, so it gets the default grad that Natalia points out. Maybe one should just add requires grad to the Tensor jit type...

**Comment by apaszke** [comment]
Ohh I see. Well, the problem with `requires_grad` propagation is that we can't do it perfectly (e.g. one if branch emits a tensor with requires_grad, the other without), so we will always return an overapproximation. I think that unfortunately there's no good way to fix that.

**Comment by t-vi** [comment]
I think it's related to having Optional only be with Tensor or so and that not having a calculated requires_grad.

**Comment by t-vi** [comment]
For my particlular problem #18360 solves this.

**Comment by t-vi** [comment]
#18360 has been reverted, so this happens again.

**Comment by t-vi** [comment]
With the specialization #18407, this is gone for optional.
